runtime/src/account_rent_state.rs

```diff
@@ -0,0 +1,165 @@
+use {
+    crate::bank::TransactionAccountRefCell,
+    enum_iterator::IntoEnumIterator,
+    solana_sdk::{account::ReadableAccount, native_loader, rent::Rent, sysvar},
+};
+
+#[derive(Debug, PartialEq, IntoEnumIterator)]
+pub(crate) enum RentState {
+    Uninitialized, // account.lamports == 0
+    NativeOrSysvar,
+    EmptyData,  // account.data.len == 0
+    RentPaying, // 0 < account.lamports < rent-exempt-minimum for non-zero data size
+    RentExempt, // account.lamports >= rent-exempt-minimum for non-zero data size
+}
+
+impl RentState {
+    pub(crate) fn from_account(account: &TransactionAccountRefCell, rent: &Rent) -> Self {
+        let account = account.1.borrow();
+        if account.lamports() == 0 {
+            Self::Uninitialized
+        } else if native_loader::check_id(account.owner()) || sysvar::is_sysvar_id(account.owner())
+        {
+            Self::NativeOrSysvar
+        } else if account.data().is_empty() {
+            Self::EmptyData
+        } else if !rent.is_exempt(account.lamports(), account.data().len()) {
+            Self::RentPaying
+        } else {
+            Self::RentExempt
+        }
```

Comment:
Technically, we could make do without the enum, and boil this logic down to a `bool`: is RentPaying or not. I rather like how explicit and clear the variants make the logic and exceptions, but open to other opinions.

Reply:
The enum is helpful to me! I always appreciate the added type safety as well.